pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,155 @@
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(extensionsv1alpha1.Object)
+			oldExtensionObj = oldObj.(extensionsv1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
+		}
+
+		if oldExtensionObj.GetExtensionStatus().GetState() == nil {
+			enqueue(queue, newObj)
+		}
+
+		if apiequality.Semantic.DeepEqual(newExtensionObj.GetExtensionStatus().GetState().Raw, oldExtensionObj.GetExtensionStatus().GetState().Raw) {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		err := s.seedClient.Client().Get(s.context, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debug("Skipping because resoruce with kind was deleted")
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
+			return reconcile.Result{}, fmt.Errorf("invalid state. extension state for resource with kind %s in namespace %s was nil", kind, req.Namespace)
+		}
+
+		shoot, err := fromCluster(req.Namespace, s.extensionsSeedClient)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shoot.GetName(), Namespace: shoot.GetNamespace()}}
+		if _, err := controllerutil.CreateOrUpdate(s.context, s.k8sGardenClient.Client(), shootState, func() error {
+			extensionState := gardencorev1alpha1helper.GetExtensionResourceStateForKind(shootState.Spec.Extensions, kind)
+			if extensionState == nil {
+				shootState.Spec.Extensions = append(shootState.Spec.Extensions, gardencorev1alpha1.ExtensionResourceState{
+					Kind:  kind,
+					State: gardencorev1alpha1.ProviderConfig{RawExtension: *state},
+				})
+			} else {
+				extensionState.State = gardencorev1alpha1.ProviderConfig{RawExtension: *state}
+			}
+			return nil
+		}); err != nil {
+			message := fmt.Sprintf("Shoot %s extension state was NOT succcessfuly synced", kind)
+			s.log.Error(message)
+			s.recorder.Event(shoot, corev1.EventTypeNormal, "ScheduledNextSync", message)
+			return reconcile.Result{RequeueAfter: time.Minute}, err
+		}
+
+		message := fmt.Sprintf("Shoot %s extension state was succcessfuly synced", kind)
+		s.log.Info(message)
+		s.recorder.Event(shoot, corev1.EventTypeNormal, "ScheduledNextSync", message)
+		return reconcile.Result{RequeueAfter: time.Minute}, nil
```

Comment:
No reason to requeue if extension was successfully synced as we only want to sync on extension resource state changes.